src_c/image.c

```diff
@@ -753,6 +753,87 @@ image_tostring(PyObject *self, PyObject *arg)
         }
         pgSurface_Unlock(surfobj);
     }
+    else if (!strcmp(format, "BGRA")) {
+        hascolorkey = 0;
+
+        string =
+            PyBytes_FromStringAndSize(NULL, (Py_ssize_t)surf->w * surf->h * 4);
+        if (!string)
+            return NULL;
+        PyBytes_AsStringAndSize(string, &data, &len);
+
+        pgSurface_Lock(surfobj);
+        switch (surf->format->BytesPerPixel) {
+            case 1:
+                for (h = 0; h < surf->h; ++h) {
+                    Uint8 *ptr = (Uint8 *)DATAROW(surf->pixels, h, surf->pitch,
+                                                  surf->h, flipped);
+                    for (w = 0; w < surf->w; ++w) {
+                        color = *ptr++;
+                        data[2] = (char)surf->format->palette->colors[color].r;
+                        data[1] = (char)surf->format->palette->colors[color].g;
+                        data[0] = (char)surf->format->palette->colors[color].b;
+                        data[3] = (char)255;
+                        data += 4;
+                    }
+                }
+                break;
+            case 2:
+                for (h = 0; h < surf->h; ++h) {
+                    Uint16 *ptr = (Uint16 *)DATAROW(
+                        surf->pixels, h, surf->pitch, surf->h, flipped);
+                    for (w = 0; w < surf->w; ++w) {
+                        color = *ptr++;
+                        data[2] = (char)(((color & Rmask) >> Rshift) << Rloss);
+                        data[1] = (char)(((color & Gmask) >> Gshift) << Gloss);
+                        data[0] = (char)(((color & Bmask) >> Bshift) << Bloss);
+                        data[3] = (char)(Amask ? (((color & Amask) >> Ashift)
+                                                  << Aloss)
+                                               : 255);
+                        data += 4;
+                    }
+                }
+                break;
+            case 3:
+                for (h = 0; h < surf->h; ++h) {
+                    Uint8 *ptr = (Uint8 *)DATAROW(surf->pixels, h, surf->pitch,
+                                                  surf->h, flipped);
+                    for (w = 0; w < surf->w; ++w) {
+#if SDL_BYTEORDER == SDL_LIL_ENDIAN
+                        color = ptr[0] + (ptr[1] << 8) + (ptr[2] << 16);
+#else
+                        color = ptr[2] + (ptr[1] << 8) + (ptr[0] << 16);
+#endif
+                        ptr += 3;
+                        data[2] = (char)(((color & Rmask) >> Rshift) << Rloss);
+                        data[1] = (char)(((color & Gmask) >> Gshift) << Gloss);
+                        data[0] = (char)(((color & Bmask) >> Bshift) << Bloss);
+                        data[3] = (char)(Amask ? (((color & Amask) >> Ashift)
+                                                  << Aloss)
+                                               : 255);
+                        data += 4;
```

Comment:
You're doing this right by copying the other ones, but I've just realized that it might be more performant to do the data stores and increments in one operation, that might generate faster code.

```
*data++ = (char)(((color & Bmask) >> Bshift) << Bloss);
*data++ = (char)(((color & Gmask) >> Gshift) << Gloss);
*data++ = (char)(((color & Rmask) >> Rshift) << Rloss);
*data++ = (char)(Amask ? (((color & Amask) >> Ashift) << Aloss) : 255);
```

Maybe not something for this PR, but it would interesting to test this.

Reply:
I mention here [later PR](https://github.com/pygame/pygame/issues/2972#issuecomment-1193689297) that I couldn't directly use `tostring_surf_32bpp()`, which can be faster. 
It can be adapted to the case of BGRA.

Reply:
I did see that, but that seems complicated. Doing the data incrementing like this could potentially be a speed up for other bit depths though, besides 32bpp.

Reply:
I can benchmark it here and see what difference it makes, at least, to my local setup.
Definitely worth exploring.
Of course, we have all seen optimisations that inexplicably make things slower on another platform/s.

Like you say, the indexing approach is used in many critical loops in `image.c` .

Reply:
Just to be clear, this hunch of mine is not based on the access cost, I think the incrementing may be faster this way.

When I took a class learning 8 bit assembly there were specific instructions for loading and storing that would also increment the location pointer by 1 at the same time, there may be some similar effect through pipelining/x86 instructions/compiler magic.

Reply:
@Starbuck5 

I'm not sure if this is the best place to continue the discussion on optimisation or create a new thread in the forum?
___________________

**Auto Increment addressing mode** - one op does access and increment (or decrement) at once,
[ref](http://www.cs.iit.edu/~cs561/cs350/addressing/addressing.html)

But not on Intel CPUs,
[stackoverflow](https://stackoverflow.com/questions/49034893/post-increment-and-pre-decrement-addressing-modes-on-intel-80386)
___________________

I put together a quick benchmark to test attempts at speedup,
As an example, I tested the current indexing method vs pointer incrementing,

```
#include <sys/time.h>
#include <stdio.h>
#include <stdlib.h>

typedef unsigned char Uint8;

typedef struct RGBA {
  Uint8 r;
  Uint8 g;
  Uint8 b;
  Uint8 a;
} RGBA_t;

void init_buffer(RGBA_t *b, size_t width, size_t height) {
  size_t i = 0;
  for (size_t h = 0; h < height; ++h) {
    for (size_t w = 0; w < width; ++w) {
      b[i].r = i % 256;
      b[i].g = (i+1) % 256;
      b[i].b = (i+2) % 256;
      b[i].a = (i+3) % 256;
      i++;
    }
  }
}

void print_buffer(RGBA_t *b, size_t width, size_t height) {
  size_t i = 0;
  for (size_t h = 0; h < height; ++h) {
    for (size_t w = 0; w < width; ++w) {
      printf("(%d,%d,%d,%d) ", b[i].r, b[i].g, b[i].b, b[i].a);
      i++;
    }
  printf("\n");
  }
}

void process_1(RGBA_t *b, size_t width, size_t height) {
  RGBA_t *ptr = b;
  Uint8 *data = (Uint8 *)b;
  for (size_t h = 0; h < height; ++h) {
    for (size_t w = 0; w < width; ++w) {
      RGBA_t color = *ptr++;
      data[0] = color.b;
      data[1] = color.g;
      data[2] = color.r;
      data[3] = color.a;
      data += 4;
    }
  }
}

void process_2(RGBA_t *b, size_t width, size_t height) {
  RGBA_t *ptr = b;
  Uint8 *data = (Uint8 *)b;
  for (size_t h = 0; h < height; ++h) {
    for (size_t w = 0; w < width; ++w) {
      RGBA_t color = *ptr++;
      *data++ = color.a;
      *data++ = color.r;
      *data++ = color.g;
      *data++ = color.b;
    }
  }
}

int main() {
  size_t width = 10000;
  size_t height = 10000;
  
  long seconds, microseconds, elapsed;
  struct timeval begin, end;
  
  RGBA_t *buf = calloc(width*height, sizeof(RGBA_t));
  
  // initialize RGBA buffer
  init_buffer(buf, width, height);
  // print_buffer(buf, width, height);
  
  // times process_1
  gettimeofday(&begin, 0);
  process_1(buf, width, height);
  gettimeofday(&end, 0);
  
  seconds = end.tv_sec - begin.tv_sec;
  microseconds = end.tv_usec - begin.tv_usec;
  elapsed = seconds*1e6 + microseconds;
  printf("%.3f ms | process_1\n", elapsed/1000.0);
  // print_buffer(buf, width, height);
  
  // time process_2
  gettimeofday(&begin, 0);
  process_2(buf, width, height);
  gettimeofday(&end, 0);
  
  seconds = end.tv_sec - begin.tv_sec;
  microseconds = end.tv_usec - begin.tv_usec;
  elapsed = seconds*1e6 + microseconds;
  printf("%.3f ms | process_2\n", elapsed/1000.0);
  // print_buffer(buf, width, height);
  
  
  free(buf);
  return 0;
}
```

output is typically,
```
59.981 ms | process_1
91.146 ms | process_2
```
using gcc (Rev3, Built by MSYS2 project) 12.1.0
cflags  = -std=c11 -Wall -O2 -Wextra

Indicating current indexing method is faster?
However,  it doesn't mean much as it is dependent on so many variables

e.g. I changed the optimisation flag from O2 to O3 and the result is typically,

```
104.040 ms | process_1
42.692 ms | process_2
```
Indicating pointer increment method can be faster :)
_______________

What about MSVC?
Microsoft doesn't have gcc's gettimeofday.
Using code from here - [stackoverflow - Equivalent of gettimeday() for Windows](https://stackoverflow.com/a/59359900/243662)

typically,
```
111.197 ms | process_1
112.680 ms | process_2

```

No difference!
using,
cl.exe - Microsoft (R) C/C++ Optimizing Compiler Version 19.29.30146 for x64
optimisation flag **/O2** (maximum speed)

_______________

The point I'm making is optimisation is a tricky business :)
_______________

An area worth exploring is only assigning those bytes that are different, between formats,

e.g. from inner loop for 4 byte color, we have,
```
      data[0] = color.b;
      data[1] = color.g;
      data[2] = color.r;
      data[3] = color.a;
      data += 4;
```

If the original data is RGBA and we are going to BGRA, we can do a datarow memory copy in the outer loop and then only swap R & B in the inner loop.

I will explore this.

Reply:
@Starbuck5 

I'm not sure if this is the best place to continue the discussion on optimisation or create a new thread in the forum?
___________________

**Auto Increment addressing mode** - one op does access and increment (or decrement) at once,
[ref](http://www.cs.iit.edu/~cs561/cs350/addressing/addressing.html)

But not on Intel CPUs,
[stackoverflow](https://stackoverflow.com/questions/49034893/post-increment-and-pre-decrement-addressing-modes-on-intel-80386)
___________________

I put together a quick benchmark to test attempts at speedup,
As an example, I tested the current indexing method vs pointer incrementing,

```
#include <sys/time.h>
#include <stdio.h>
#include <stdlib.h>

typedef unsigned char Uint8;

typedef struct RGBA {
  Uint8 r;
  Uint8 g;
  Uint8 b;
  Uint8 a;
} RGBA_t;

void init_buffer(RGBA_t *b, size_t width, size_t height) {
  size_t i = 0;
  for (size_t h = 0; h < height; ++h) {
    for (size_t w = 0; w < width; ++w) {
      b[i].r = i % 256;
      b[i].g = (i+1) % 256;
      b[i].b = (i+2) % 256;
      b[i].a = (i+3) % 256;
      i++;
    }
  }
}

void print_buffer(RGBA_t *b, size_t width, size_t height) {
  size_t i = 0;
  for (size_t h = 0; h < height; ++h) {
    for (size_t w = 0; w < width; ++w) {
      printf("(%d,%d,%d,%d) ", b[i].r, b[i].g, b[i].b, b[i].a);
      i++;
    }
  printf("\n");
  }
}

void process_1(RGBA_t *b, size_t width, size_t height) {
  RGBA_t *ptr = b;
  Uint8 *data = (Uint8 *)b;
  for (size_t h = 0; h < height; ++h) {
    for (size_t w = 0; w < width; ++w) {
      RGBA_t color = *ptr++;
      data[0] = color.b;
      data[1] = color.g;
      data[2] = color.r;
      data[3] = color.a;
      data += 4;
    }
  }
}

void process_2(RGBA_t *b, size_t width, size_t height) {
  RGBA_t *ptr = b;
  Uint8 *data = (Uint8 *)b;
  for (size_t h = 0; h < height; ++h) {
    for (size_t w = 0; w < width; ++w) {
      RGBA_t color = *ptr++;
      *data++ = color.a;
      *data++ = color.r;
      *data++ = color.g;
      *data++ = color.b;
    }
  }
}

int main() {
  size_t width = 10000;
  size_t height = 10000;
  
  long seconds, microseconds, elapsed;
  struct timeval begin, end;
  
  RGBA_t *buf = calloc(width*height, sizeof(RGBA_t));
  
  // initialize RGBA buffer
  init_buffer(buf, width, height);
  // print_buffer(buf, width, height);
  
  // times process_1
  gettimeofday(&begin, 0);
  process_1(buf, width, height);
  gettimeofday(&end, 0);
  
  seconds = end.tv_sec - begin.tv_sec;
  microseconds = end.tv_usec - begin.tv_usec;
  elapsed = seconds*1e6 + microseconds;
  printf("%.3f ms | process_1\n", elapsed/1000.0);
  // print_buffer(buf, width, height);
  
  // time process_2
  gettimeofday(&begin, 0);
  process_2(buf, width, height);
  gettimeofday(&end, 0);
  
  seconds = end.tv_sec - begin.tv_sec;
  microseconds = end.tv_usec - begin.tv_usec;
  elapsed = seconds*1e6 + microseconds;
  printf("%.3f ms | process_2\n", elapsed/1000.0);
  // print_buffer(buf, width, height);
  
  
  free(buf);
  return 0;
}
```

output is typically,
```
59.981 ms | process_1
91.146 ms | process_2
```
using gcc (Rev3, Built by MSYS2 project) 12.1.0
cflags  = -std=c11 -Wall -O2 -Wextra

Indicating current indexing method is faster?
However,  it doesn't mean much as it is dependent on so many variables

e.g. I changed the optimisation flag from O2 to O3 and the result is typically,

```
104.040 ms | process_1
42.692 ms | process_2
```
Indicating pointer increment method can be faster :)
_______________

What about MSVC?
Microsoft doesn't have gcc's gettimeofday.
Using code from here - [stackoverflow - Equivalent of gettimeday() for Windows](https://stackoverflow.com/a/59359900/243662)

typically,
```
111.197 ms | process_1
112.680 ms | process_2

```

No difference!
using,
cl.exe - Microsoft (R) C/C++ Optimizing Compiler Version 19.29.30146 for x64
optimisation flag **/O2** (maximum speed)

_______________

The point I'm making is optimisation is a tricky business :)

Reply:
> The point I'm making is optimisation is a tricky business :)

That's the truth.

I was just wondering if there was an easy win there, and I guess not.